### PR TITLE
Issue #17: Implemented eth_getBlockByNumber()

### PR DESCRIPTION
**Pull request by yurenju**

### CODE REVIEW

File: tests/rpc.js [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+  }
+
+  t.false(blockchain.getBlock.called)
+  manager.execute(req, (err, block) => {
+    t.error(err)
+    t.true(blockchain.getBlock.called)
+    t.equal(blockchain.getBlock.firstCall.args[0], 1)
+    t.end()
+  })
+})
```
Comment (holgerd77) [inline tests/rpc.js]: So great that you already added some framing for test execution, really like this ``sinon`` library, didn't know that before.

I think our general reference for RPC execution should  be Geth behavior, this is also really practical to test running the JS client and Geth with RPC and different port side by side and then have a look at the results.

I've done this with JS ``8545`` and ``Geth`` ``8546`` and get divergent behavior e.g. for:

```shell
curl -X POST --data '{"jsonrpc":"2.0","method":"eth_getBlockByNumber","params":["0x1", true],"id":1}' http://127.0.0.1:8545
{...JSON Result...}
curl -X POST --data '{"jsonrpc":"2.0","method":"eth_getBlockByNumber","params":["0x1", true],"id":1}' http://127.0.0.1:8546
invalid content type, only application/json is supported
```

So Geth requires a content header ``-H "Content-Type: application/json"`` which I think we should then require as well.

Also things like leading zeros (0x01) differ:
```shell
curl -X POST --data '{"jsonrpc":"2.0","method":"eth_getBlockByNumber","params":["0x01", true],"id":1}' -H "Content-Type: application/json" http://127.0.0.1:8545|jq
{...JSON Result...}
curl -X POST --data '{"jsonrpc":"2.0","method":"eth_getBlockByNumber","params":["0x01", true],"id":1}' -H "Content-Type: application/json" http://127.0.0.1:8546|jq
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   197  100   117  100    80  73079  49968 --:--:-- --:--:-- --:--:--  114k
{
  "jsonrpc": "2.0",
  "id": 1,
  "error": {
    "code": -32602,
    "message": "invalid argument 0: hex number with leading zero digits"
  }
}
```

Can you test some basic edge cases (e.g. not-existing/downloaded block) and see how things behave and then adopt the code base to make sure that the output is the same?

Comment (holgerd77) [reply tests/rpc.js]: Hi @yurenju, I would probably merge https://github.com/ethereumjs/ethereumjs-client/pull/29 before this PR and then ask you to rebase on top, if you see any problems here let me know.

Comment (yurenju) [reply tests/rpc.js]: @holgerd77 sure go ahead! I may update my pr in two days when I get more free time :D, will rebase after #29 is landed.

File: lib/index.js [hunk 2]
```diff
@@ -22,6 +23,11 @@ function runClient () {
   var nm = new EthNetworkManager(PRIVATE_KEY, config)
 
   var cm = new ChainManager(config, nm) // eslint-disable-line
+
+  if (config.rpc) {
+    const rpc = new RPCServer(cm.getBlockchain())
+    rpc.listen(config.rpcport)
```
[the remark this answers is not in the document]
Comment (yurenju) [reply lib/index.js]: I would like to follow convention from other modules to have an argument `config` into this module and save `_config` and `_logger` into RPCServer and RPCManager

File: lib/rpc/rpc-manager.js [hunk 3]
```diff
@@ -0,0 +1,21 @@
+const modules = require('./modules')
+
+class RPCServer {
```
[the remark this answers is not in the document]
Comment (yurenju) [reply lib/rpc/rpc-manager.js]: I forgot to rename this one, nice catch!

File: lib/rpc/modules/eth.js [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+  }
+
+  getBlockByNumber (params, cb) {
+    let [blockNumber] = params
+    blockNumber = Number.parseInt(blockNumber, 16)
+    this._chain.getBlock(blockNumber, (err, block) => {
+      cb(err, block.toJSON(true))
+    })
+  }
+}
```
[the remark this answers is not in the document]
Comment (yurenju) [reply lib/rpc/modules/eth.js]: Which part do you want to replicate from `ganache-core`? 

took a look I feel we should expand argument like `function(block_number, include_full_transactions, callback)` instead of `(params, cb)` to make it more clear. 

I'm not familiar with `ganache-core` codebase, so I don't have any opinion for that yet, but I agree we should analyze `ganache-core` deeper to get a better idea to implement rpc.

File: tests/rpc.js (hunk 1 above)
Comment (yurenju) [reply tests/rpc.js]: and thanks for testing with js client and geth side by side, I will add some basic edge cases!

Comment (yurenju) [reply tests/rpc.js]: I feel the way I handled rpc is too simple since we need to fit JSON-RPC spec. there are few Node.js packages for JSON-RPC and seems jayson is a well-maintain solution.

so on next commit, I will introduce jayson.

https://github.com/tedeh/jayson

**Comment by yurenju** [comment]
Just heads up I didn't finish my commit this weekend but already added some edge test case for invalid parameters, only accept content-type as `application/json`, etc.

I pushed my WIP commit to the previous PR to show the progress. hopefully I can finish this commit this week.

**Comment by holgerd77** [comment]
Ah, already running this with ``jayson``, cool! 😄 Thanks for re-thinking this so constantly, this will be really really valuable if we have a solid structure here.

I now merged https://github.com/ethereumjs/ethereumjs-client/pull/29, can you rebase your PR on top?

**Comment by yurenju** [comment]
just rebased it. the last commit introduces some changes:

* test cases for:
  * http status code should be 415 if `Content-Type` is not `application/json`
  * returns error if method does not exist
  * returns error if parameter is invalid, include hex and boolean checker
* jayson: JSON-RPC module implementation with batch request and correct status code 
* middleware for validation

let me know if you found anything can be improved :D

**Comment by holgerd77** [comment]
@yurenju Thanks again, looking into it right now (think I will learn a lot reviewing your work...).

File: lib/cliParser.js [hunk 5]
```diff
@@ -11,6 +11,17 @@ var parser = require('yargs')
...
+    describe: 'Enable the JSON-RPC server'
+  })
+  .option('rpcport', {
+    describe: 'HTTP-RPC server listening port',
+    default: 8545
+  })
+  .option('rpcaddr', {
+    describe: 'HTTP-RPC server listening interface',
+    default: 'localhost'
+  })
```
Comment (holgerd77) [inline lib/cliParser.js]: Nice that you also added the ``rpcaddr`` now.

File: lib/index.js [hunk 6]
```diff
@@ -31,6 +34,14 @@ function runClient () {
   })
 
   var cm = new ChainManager(config, nm, bc) // eslint-disable-line
+
+  if (config.rpc) {
+    const { rpcport, rpcaddr } = config
+    const manager = new RPCManager(cm.getBlockchain(), config)
+    const server = jayson.server(manager.getMethods())
+    config.logger.info(`HTTP endpoint opened: http://${rpcaddr}:${rpcport}`)
+    server.http().listen(rpcport)
```
Comment (holgerd77) [inline lib/index.js]: Was a bit sceptical about the naming but wasn't aware that ``consts`` in an ``if`` clause remain local not trickling in the global space (tested this), so this should be ok.

File: lib/rpc/modules/eth.js [hunk 7]
```diff
@@ -0,0 +1,25 @@
+const { middleware, validators } = require('../validation')
+
+class Eth {
+  constructor (chain) {
+    this._chain = chain
+
+    this.getBlockByNumber = middleware(this.getBlockByNumber.bind(this), 2,
+      [validators.hex, validators.bool])
+  }
```
Comment (holgerd77) [inline lib/rpc/modules/eth.js]: This is a really super-nice middleware structure.

File: lib/rpc/validation.js [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+module.exports = {
+  /**
+   * middleware for parameters validation
+   * @param {Function} method            function to add middleware
+   * @param {number} requiredParamsCount required parameters count
+   * @param {Function[]} validators      array of validator
+   */
+  middleware (method, requiredParamsCount, validators) {
+    return function (params, cb) {
+      if (params.length < requiredParamsCount) {
```
Comment (holgerd77) [inline lib/rpc/validation.js]: Wonder if this should be a ``!==`` comparison and a bit more general error msg, but we can leave this for now. Another possibility would also to make an extra check for ``>``, this is maybe even more user-friendly for the caller.

File: tests/rpc/eth/getBlockByNumber.js [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+  const server = startRPC(manager.getMethods())
+
+  const req = {
+    jsonrpc: '2.0',
+    method: 'eth_getBlockByNumber',
+    params: ['0x1', true],
+    id: 1
+  }
+
+  request(server)
```
Comment (holgerd77) [inline tests/rpc/eth/getBlockByNumber.js]: Nice setup with this ``supertest`` library - didn't know that yet - really sleak.

**Comment by holgerd77** [comment]
For testing this locally:

```shell
curl -X POST --data '{"jsonrpc":"2.0","method":"eth_getBlockByNumber","params":["0x1b4", true],"id":1}' -H "Content-Type: application/json" http://127.0.0.1:8545|jq
```

(``jq`` is for (optional) formatting and has to be installed manually before)

Result:

```json
{
  "jsonrpc": "2.0",
  "id": 1,
  "result": {
    "header": {
      "parentHash": "0xe99e022112df268087ea7eafaf4790497fd21dbeeb6bd7a1721df161a6657a54",
      "uncleHash": "0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347",
      "coinbase": "0xbb7b8287f3f0a933474a79eae42cbca977791171",
      "stateRoot": "0xddc8b0234c2e0cad087c8b389aa7ef01f7d79b2570bccb77ce48648aa61c904d",
      "transactionsTrie": "0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421",
      "receiptTrie": "0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421",
      "bloom": "0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
      "difficulty": "0x04ea3f27bc",
      "number": "0x01b4",
      "gasLimit": "0x1388",
      "gasUsed": "0x",
      "timestamp": "0x55ba467c",
      "extraData": "0x476574682f4c5649562f76312e302e302f6c696e75782f676f312e342e32",
      "mixHash": "0x4fffe9ae21f1c9e15207b1f472d5bbdd68c9595d461666602f2be20daf5e7843",
      "nonce": "0x689056015818adbe"
    },
    "transactions": [],
    "uncleHeaders": []
  }
}
```

File: lib/rpc/validation.js (hunk 8 above)
Comment (yurenju) [reply lib/rpc/validation.js]: I did this because want to align with go-ethereum behaviors, but you are right we can handle this better than geth :D

**Comment by Nathapon8899** [comment]
#17
